design/ironic-standalone-operator.md

```diff
@@ -0,0 +1,369 @@
+- `gatewayAddress` - IP address of the default gateway (if necessary)
+
+Providing a non-nil `dhcp` value enables dnsmasq.
+
+### Risks and Mitigations
+
+Our reliance on host networking means that it's not trivial to have several
+Ironic installations on the same cluster. Each would need to use different
+ports to avoid conflicts. Even without host networking, having several dnsmasq
+instances on the same network is not going to work without some sort of
+coordination between them.
+
+### Work Items
+
+General enablement:
+
+- Add the operator to the development environment.
+- Add an optional flag either to metal3-dev-env or to BMO e2e tests (TBD) that
+  uses ironic-standalone-operator instead of the Kustomize scripts in BMO.
+- Create and run CI jobs (integration or e2e - depending on the previous work
+  item) on the operator.
+
+HA:
+
+- Implement the boot configuration API in Ironic (dependency).
+- Start generating a private CA for JSON RPC.
+- Enable the HA architecture.
+- Adjust ironic-image to enable updates without wiping the database (see
+  below).
+
+### Dependencies
+
+None for the core operator.
+
+The HA approach will require [boot configuration API][boot config].
+
+### Test Plan
+
+The new operator will become the primary way to install Ironic. As such, it
+will be tested in various CI jobs.
+
+### Upgrade / Downgrade Strategy
+
+By default, the operator will be tightly coupled with the version of Ironic
+(and, eventually, IPA) that it installs. A release of the operator will follow
+each release of ironic-image, and they release branches will match. The `main`
+branch will continue following the latest container image.
+
+After each full reconciliation, the operator will store the version of Ironic
+it has just installed in the `status`. In the future, this will allow to apply
+any logic on upgrade. However, we will try to keep ironic-image self-upgrading
+for the sake of users that do not use ironic-standalone-operator.
+
+To accommodate downstream modifications (like in OpenShift), it will be
+possible to modify all images, as well as the installed version, via
+environment variables.
+
+#### Database Migrations
+
+Having MariaDB as a separate container also poses a new challenge for Metal3
+since now Ironic will sometimes start with the database already populated
+rather than a clean one (as is the case for SQLite). To accommodate this:
+
+- We'll create a new container entrypoint that will run the *online data
+  migrations* for Ironic while the service is already running (a part of
+  the upgrade process that we've ignored so far).
+- We'll probably need to upgrade the database schema separately from Ironic to
+  avoid running it 3+ times in parallel. Maybe it will take a form of a *Job*.
+- BMO will need to be updated to handle more cases of unexpected provision
+  state. E.g. what needs to be done when the BMH is *inspecting* but the node
+  is found in a completely wrong state like `active` or `clean wait`.
```

Comment:
There is an additional complexity to consider if we're going to support mariadb combined with persistent volumes, when the mariadb version changes you either have to ensure clean shutdown of the container or enable some kind of backup/restore workflow due to the mariadb checks around versioning for the redo log - [see here](https://github.com/suse-edge/charts/pull/135#issuecomment-2236635629) for example.

This can likely be resolved with with pod lifecycle hooks, or perhaps some kind of job pre/post upgrade but it's additional complexity which I don't think we currently consider in the community deployment examples.

Reply:
Very good point, thank you!